pin/gc/gc.go

```diff
@@ -0,0 +1,112 @@
+package gc
+
+import (
+	bstore "github.com/ipfs/go-ipfs/blocks/blockstore"
+	key "github.com/ipfs/go-ipfs/blocks/key"
+	bserv "github.com/ipfs/go-ipfs/blockservice"
+	offline "github.com/ipfs/go-ipfs/exchange/offline"
+	dag "github.com/ipfs/go-ipfs/merkledag"
+	pin "github.com/ipfs/go-ipfs/pin"
+
+	context "github.com/ipfs/go-ipfs/Godeps/_workspace/src/golang.org/x/net/context"
+	eventlog "github.com/ipfs/go-ipfs/thirdparty/eventlog"
+)
+
+var log = eventlog.Logger("gc")
+
+// GC performs a mark and sweep garbage collection of the blocks in the blockstore
+// first, it creates a 'marked' set and adds to it the following:
+// - all recursively pinned blocks, plus all of their descendants (recursively)
+// - all directly pinned blocks
+// - all blocks utilized internally by the pinner
+//
+// The routine then iterates over every block in the blockstore and
+// deletes any block that is not found in the marked set.
+func GC(ctx context.Context, bs bstore.GCBlockstore, pn pin.Pinner) (<-chan key.Key, error) {
+	unlock := bs.GCLock()
+	defer unlock()
+
+	bsrv, err := bserv.New(bs, offline.Exchange(bs))
+	if err != nil {
+		return nil, err
+	}
+	ds := dag.NewDAGService(bsrv)
+
+	gcs, err := ColoredSet(pn, ds)
+	if err != nil {
+		return nil, err
+	}
+
+	keychan, err := bs.AllKeysChan(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	output := make(chan key.Key)
+	go func() {
+		defer close(output)
+		for {
+			select {
+			case k, ok := <-keychan:
+				if !ok {
+					return
+				}
+				if !gcs.Has(k) {
+					err := bs.DeleteBlock(k)
+					if err != nil {
+						log.Debugf("Error removing key from blockstore: %s", err)
+						return
+					}
+					select {
+					case output <- k:
+					case <-ctx.Done():
+						return
+					}
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return output, nil
+}
+
+func Descendants(ds dag.DAGService, set key.KeySet, roots []key.Key) error {
+	for _, k := range roots {
+		set.Add(k)
+		nd, err := ds.Get(context.Background(), k)
+		if err != nil {
+			return err
+		}
+
+		// EnumerateChildren recursively walks the dag and adds the keys to the given set
+		err = dag.EnumerateChildren(context.Background(), ds, nd, set)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func ColoredSet(pn pin.Pinner, ds dag.DAGService) (key.KeySet, error) {
+	// KeySet currently implemented in memory, in the future, may be bloom filter or
+	// disk backed to conserve memory.
```

Comment:
if you have 100GB disk space to devote to ipfs, and all 100GB of that is pinned, i'm sure you have more than 15MB of ram to devote to the task of GCing. Similarly, i've never seen a machine with 10TB of storage that doesnt have at least 2 GB of ram. It's a definite TODO for us, but right now i think that we're fine pushing it back a ways. Other things are going to kill memory usage before GC will at that scale.